### ISSUE TITLE
Loss weighting behavior will give inconsistent "upweighting" between batches

### ISSUE THREAD

**Issue by yair-schiff** [songlab-cal/gpn#23]
Hi there!

I had a quick question about the Cross-Entropy loss calculation [here](https://github.com/songlab-cal/gpn/blob/05b23c54c572813810c094b31031901f7109575b/gpn/model.py#L74-L82):
```
    elif labels is not None and loss_weight is not None:
        loss_fct = CrossEntropyLoss(reduction="none")
        labels = labels.view(-1)
        loss = loss_fct(
            logits.view(-1, vocab_size), labels
        )  # what if we first exclude the ones with -100??
        loss_weight = loss_weight.view(-1)
        loss_weight[labels == -100] = 0.0
        loss = (loss * loss_weight / loss_weight.sum()).sum()
```

While I understand the goal is to weight tokens differently, i.e. place more weight on "upper" case base pairs that are 'not repetead', I think the way the code is written leads to slightly unintuitive behavior. That is, on any given sequence / batch the weight of the token will be determined by the other masked tokens.

Take a simple example `aaCTgg`, where 2 tokens will be masked. If an upper and a lower case token are masked then the upper case token will correctly have a higher weighting in the loss. However if 2 upper case tokens are masked or if 2 lower case tokens are masked they will then have equal weight and therefore contribute to the loss the exact same. This seems counter to the desired behavior.

I think on aggregate this is not a 'big deal' since on average there will be a mix of up/down weighted tokens in any given sequence/batch but I wanted to check if perhaps this is slightly inconsistent with the desired behavior.

**Comment by gonzalobenegas** [comment]
Hello! I'd also think that this might not be a big deal when using a large enough batch. I appreciate your thoughts, how would you reformulate this?

**Comment by yair-schiff** [comment]
I agree that it's a tricky question on how to do this right. Perhaps forgoing the weighted CE being an actual average, i.e., remove the loss weight normalization?
```
loss = (loss * loss_weight).sum()
```

This might make it a bit less intuitive as to how to compare GPN pre-train loss to other models though.

What are your thoughts on this approach?
